File: August/Week2/Day2/Submissions/rottenOranges.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var orangesRotting = function (grid) {
    const rNum = grid.length, cNum = grid[0].length;
    const rD = [1, -1, 0, 0], cD = [0, 0, 1, -1];
    let queue = [], minute = 0, fresh = 0;

    for (let r = 0; r < rNum; r++) {
        for (let c = 0; c < cNum; c++) {
            if (grid[r][c] === 2) queue.push([r, c])
            else if (grid[r][c] === 1) fresh++;
        }
    }

    while (fresh && queue.length) {
        const next = [];
        minute++;

        for (let [r, c] of queue) {

            for (let i = 0; i < rD.length; i++) {
                const nR = r + rD[i], nC = c + cD[i];

                if (nR < rNum && nR >= 0 &&
                    nC < cNum && nC >= 0 &&
                    grid[nR][nC] === 1) {
                    fresh--;
                    grid[nR][nC] = 2;
                    next.push([nR, nC]);
                }
            }
        }
        queue = next;
    }
    return !fresh ? minute : -1;
};

File: August/Week2/Day2/workArea/rottenOranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        # print(grid)
        '''minuteCounter = 0
        for i in range(len(grid)):
            # print(i)
            for j in range(len(grid)):
                # print(grid[i][j])
                if grid[i][j] == 2:
                    # minuteCounter += 1
                    # print('Hey', minuteCounter)
                    if (i-1) > 0 or (j-1) > 0 or (i+1) > 0 and (i+1) < len(grid) and (j+1) > 0 and (j+1) < len(grid):
                        grid[i-1][j] = 2
                        grid[i+1][j] = 2
                        grid[i][j-1] = 2
                        grid[i][j+1] = 2

                        minuteCounter += 1
                    
                else:
                    continue
                    
        return minuteCounter'''
        m, n, queue, fresh = len(grid), len(grid[0]), deque(), 0
        for i,j in product(range(m), range(n)):
            if grid[i][j] == 2: queue.append((i,j))
            if grid[i][j] == 1: fresh += 1
        dirs = [[1,0],[-1,0],[0,1],[0,-1]]
        levels = 0
        
        while queue:
            levels += 1
            for _ in range(len(queue)):
                x, y = queue.popleft()
                for dx, dy in dirs:
                    if 0<=x+dx<m and 0<=y+dy<n and grid[x+dx][y+dy] == 1:
                        fresh -= 1
                        grid[x+dx][y+dy] = 2
                        queue.append((x+dx, y+dy))
                        
        return -1 if fresh != 0 else max(levels-1, 0)
        